### PR TITLE
[NFC][Utils] Extract CloneFunctionMetadataInto from CloneFunctionInto

### PR DESCRIPTION
**Pull request by artempyanykh**
Stacked PRs:
 * #118630
 * #118629
 * #118628
 * #118627
 * #118625
 * #118624
 * __->__#118623

--- --- ---

### [NFC][Utils] Extract CloneFunctionMetadataInto from CloneFunctionInto

Summary:
The new API expects the caller to populate the VMap. We need it this way
for a subsequent change around coroutine cloning.

Test Plan:
ninja check-llvm-unit check-llvm

### CODE REVIEW

File: llvm/include/llvm/Transforms/Utils/Cloning.h [hunk 1]
```diff
@@ -182,6 +182,18 @@ void CloneFunctionAttributesInto(Function *NewFunc, const Function *OldFunc,
...
                                  ValueMaterializer *Materializer = nullptr);
 
+/// Clone OldFunc's metadata into NewFunc.
+///
+/// The caller is expected to populate \p VMap beforehand and set an appropriate
+/// \p RemapFlag.
+///
+/// NOTE: This function doesn't clone !llvm.dbg.cu when cloning into a different
+/// module. Use CloneFunctionInto for that behavior.
+void CloneFunctionMetadataInto(Function *NewFunc, const Function *OldFunc,
```
Comment (felipepiovezan) [inline llvm/include/llvm/Transforms/Utils/Cloning.h:192]: Can `NewFunc` or `OldFunc` be null? If not, we should make them references

File: llvm/lib/Transforms/Utils/CloneFunction.cpp [hunk 2]
```diff
@@ -200,6 +200,22 @@ bool llvm::BuildDebugInfoMDMap(DenseMap<const Metadata *, TrackingMDRef> &MD,
   return ModuleLevelChanges;
 }
 
+void llvm::CloneFunctionMetadataInto(Function *NewFunc, const Function *OldFunc,
+                                     ValueToValueMapTy &VMap,
+                                     RemapFlags RemapFlag,
+                                     ValueMapTypeRemapper *TypeMapper,
+                                     ValueMaterializer *Materializer) {
+  // Duplicate the metadata that is attached to the cloned function.
```
Comment (felipepiovezan) [inline llvm/lib/Transforms/Utils/CloneFunction.cpp:208]: Note that, in the context of this function, there is no "cloned function", it's either `NewFunc` or `OldFunc`. 

That  said, I this comment fits better in the documentation of the header, as there is important information here that callers should be aware

File: llvm/include/llvm/Transforms/Utils/Cloning.h (hunk 1 above)
Comment (artempyanykh) [reply llvm/include/llvm/Transforms/Utils/Cloning.h]: Both can't be null. Made them references.

File: llvm/lib/Transforms/Utils/CloneFunction.cpp (hunk 2 above)
Comment (artempyanykh) [reply llvm/lib/Transforms/Utils/CloneFunction.cpp]: Thanks! Moved relevant documentation bits into Cloning.h.

**Review by felipepiovezan — APPROVED** [review]

**Pull request merged**